File: 0155_min_stack.py
# https://leetcode-cn.com/problems/min-stack/

class Solution:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.stack = []
        
    def push(self, x):
        """
        Inputs:
            x: int
        Outputs:
            None
        """
        if not self.stack:
            self.stack.append((x,x))
        else:
            self.stack.append((x,min(x,self.stack[-1][1])))
            
    def pop(self):
        """
        Inputs:
            None
        Outputs:
            None
        """
        self.stack.pop()
        
    def top(self):
        """
        Inputs:
            None
        Outputs:
            return int
        """
        return self.stack[-1][0]
    
    def getMin(self):
        """
        Inputs:
            None
        Outputs:
            return int
        """
        return self.stack[-1][1]
    
# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: Cpp/0155_min_stack.cpp
// https://leetcode.cn/problems/min-stack/submissions/

class MinStack {
public:
    MinStack() {
    }
    
    void push(int val) {
        if (len == 0) {
            stack_.push_back((vector<int>){val, val});
        }
        else {
            stack_.push_back((vector<int>){val, min(stack_[len-1][1], val)});
        }
        len = len + 1;
    }
    
    void pop() {
        stack_.pop_back();
        len = len - 1;
    }
    
    int top() {
        return stack_[len-1][0];
    }
    
    int getMin() {
        return stack_[len-1][1];
    }

private:
    vector<vector<int>> stack_;
    int len = 0;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */